src/lib/merkle_ledger/database.ml

```diff
@@ -238,6 +244,158 @@ module Make (Inputs : Inputs_intf) :
+      (* This assumes that token owners are dense in the interval
+         [default+1, last_token]. Otherwise, we should track an indirection,
+         like we do with accounts.
+      *)
+      let rec go tid =
+        let tid = Token_id.next tid in
+        if Token_id.(tid <= last_token) then
+          match get t tid with
+          | Some pk ->
+              (tid, pk) :: go tid
+          | None ->
+              go tid
+        else []
+      in
+      go Token_id.default
+  end
+
+  (** The tokens associated with each public key.
+
+      These are represented as a [Token_id.Set.t], which is represented by an
+      ordered list.
+  *)
+  module Tokens = struct
+    let build_location pk =
+      Location.build_generic
+        (Bigstring.of_string (Format.sprintf !"$tids!%{sexp: Key.t}" pk))
+
+    let serialize_kv (pk, tids) =
+      let tokens_buf = Bigstring.create (Token_id.Set.bin_size_t tids) in
+      ignore (Token_id.Set.bin_write_t tokens_buf ~pos:0 tids) ;
+      (Location.serialize (build_location pk), tokens_buf)
+
+    let get_opt mdb pk =
+      get_bin mdb (build_location pk) Token_id.Set.bin_read_t
+
+    let get mdb pk = Option.value ~default:Token_id.Set.empty (get_opt mdb pk)
+
+    let set mdb pk tids =
+      set_bin mdb (build_location pk) Token_id.Set.bin_size_t
+        Token_id.Set.bin_write_t tids
+
+    let delete mdb pk = delete_raw mdb (build_location pk)
+
+    let change_opt mdb pk ~f =
+      let old = get_opt mdb pk in
+      match (old, f old) with
+      | _, Some tids ->
+          set mdb pk tids
+      | Some _, None ->
+          delete mdb pk
+      | None, None ->
+          ()
+
+    let to_opt s = if Set.is_empty s then None else Some s
+
+    let _update_opt mdb pk ~f = change_opt mdb pk ~f:(fun x -> to_opt (f x))
+
+    let update mdb pk ~f =
+      change_opt mdb pk ~f:(fun x ->
+          to_opt @@ f (Option.value ~default:Token_id.Set.empty x) )
+
+    let add mdb pk tid = update mdb pk ~f:(fun tids -> Set.add tids tid)
+
+    let _add_several mdb pk new_tids =
+      update mdb pk ~f:(fun tids ->
+          Set.union tids (Token_id.Set.of_list new_tids) )
+
+    let add_account mdb aid =
+      add mdb (Account_id.public_key aid) (Account_id.token_id aid)
+
+    let remove mdb pk tid = update mdb pk ~f:(fun tids -> Set.remove tids tid)
+
+    let _remove_several mdb pk rem_tids =
+      update mdb pk ~f:(fun tids ->
+          Set.diff tids (Token_id.Set.of_list rem_tids) )
+
+    let remove_account mdb aid =
+      remove mdb (Account_id.public_key aid) (Account_id.token_id aid)
+
+    (** Generate a batch of database changes to add the given tokens. *)
+    let add_batch_create mdb pks_to_tokens =
+      let pks_to_all_tokens =
+        Map.filter_mapi pks_to_tokens ~f:(fun ~key:pk ~data:tokens_to_add ->
+            to_opt (Set.union (get mdb pk) tokens_to_add) )
+      in
+      Map.to_alist pks_to_all_tokens |> List.map ~f:serialize_kv
+
+    let _add_batch mdb pks_to_tokens =
```

Comment:
Are these functions merely defined so that they can be easily exposed later?

Reply:
Yes. I can remove them if it's a code smell.